examples/websockets/src/main.rs

```diff
@@ -59,58 +76,169 @@ async fn main() {
+/// Actual websocket statemachine (one will be spawned per connection)
+async fn handle_socket(mut socket: WebSocket, who: SocketAddr) {
+    //send a ping (unsupported by some browsers) just to kick things off and get a response
+    if let Ok(_) = socket.send(Message::Ping(vec![1, 2, 3])).await {
+        println!("Pinged {}...", who);
+    } else {
+        println!("Could not ping {}!", who);
+        return;
+    }
+
+    // receive single message form a client (we can either receive or send with socket)
+    // this will likely be the Pong for our Ping or a hello.
     if let Some(msg) = socket.recv().await {
         if let Ok(msg) = msg {
-            match msg {
-                Message::Text(t) => {
-                    println!("client sent str: {:?}", t);
-                }
-                Message::Binary(_) => {
-                    println!("client sent binary data");
-                }
-                Message::Ping(_) => {
-                    println!("socket ping");
-                }
-                Message::Pong(_) => {
-                    println!("socket pong");
-                }
-                Message::Close(_) => {
-                    println!("client disconnected");
-                    return;
-                }
+            if process_message(msg, who) {
+                return;
             }
         } else {
-            println!("client disconnected");
+            println!("{} abruptly disconnected", who);
             return;
         }
     }
 
-    loop {
+    // we can also send messages to client by simply sleeping,
+    // since each client gets individual statemachine. Thus, waiting for this
+    // client to finish getting his greetings does not prevent other clients form
+    // connecting to server.
+    for i in 1..5 {
         if socket
-            .send(Message::Text(String::from("Hi!")))
+            .send(Message::Text(String::from(format!("Hi {} times!", i))))
             .await
             .is_err()
         {
-            println!("client disconnected");
+            println!("client abruptly disconnected");
             return;
         }
-        tokio::time::sleep(std::time::Duration::from_secs(3)).await;
+        tokio::time::sleep(std::time::Duration::from_millis(100)).await;
+    }
+
+    // By splitting socket we can send and receive at the same time. In this example we will send
+    // unsolicited messages to client based on some sort of server's internal event (i.e .timer).
+    let (mut sender, mut receiver) = socket.split();
+
+    // Spawn a task that will push 100 messages to the client (does not matter what client does)
+    let mut send_task = tokio::spawn(async move {
+        let mut i = 20;
+        while i > 0 {
+            // In any websocket error, break loop.
+            if sender
+                .send(Message::Text(format!("{} messages left...", i)))
+                .await
+                .is_err()
+            {
+                break;
```

Comment:
```suggestion
                return;
```

That way you don't need to check `i` after the loop because you'll return from the task.

Reply:
Yes, figured a way to rewrite that without while. Looks better now. Will add that to the bigger commit.